core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1874,13 +1879,13 @@ class ReplicaManager(val config: KafkaConfig,
     readPartitionInfo.foreach { case (tp, fetchInfo) =>
       val readResult = read(tp, fetchInfo, limitBytes, minOneMessage)
       val recordBatchSize = readResult.info.records.sizeInBytes
-      // Once we read from a non-empty partition, we stop ignoring request and partition level size limits
-      if (recordBatchSize > 0)
-        minOneMessage = false
       // Because we don't know how much data will be retrieved in remote fetch yet, and we don't want to block the API call
       // to query remoteLogMetadata, assume it will fetch the max bytes size of data to avoid to exceed the "fetch.max.bytes" setting.
       val estimatedRecordBatchSize = if (recordBatchSize == 0 && readResult.info.delayedRemoteStorageFetch.isPresent)
         readResult.info.delayedRemoteStorageFetch.get.fetchMaxBytes else recordBatchSize
+      // Once we read from a non-empty partition, we stop ignoring request and partition level size limits
+      if (estimatedRecordBatchSize > 0)
+        minOneMessage = false
```

Comment:
Let us keep this PR only for the memory leak issue in remote reads. Please have another PR related to `minOneMessage` scenario.

Reply:
Opened #20706 to fix minOneMessage case. 

@showuon 
One additional change in #20706 is that remoteFetchInfos changed from HashMap to LinkedHashMap to preserve the order of partitions while sending the remote fetches. PTAL. 

Thanks for the review!